### PR TITLE
Added new algorithm, restructured the folder, and more.

### PR DESCRIPTION
**Pull request by pkofod**
A Frisch-Newton interior point algorithm has been implemented (Koenker & Portnoy 1997). It has been benchmarked (issue #4) to be quite fast. @andreasnoack has contributed with valuable optimizations of the implementation. Each algorithm has been placed in their own file, and the variance-covariance estimation has also been isolated.

I've placed @vincentarelbundock license notice in IRLS.jl and Covariance.jl.

### CODE REVIEW

**Comment by pkofod** [comment]
For reference, here are a more thorough benchmark of the 40 explanatory variables case from #4 . Improved is the version after Noacks changes where implemented.

results_time
results_mem

**Comment by pkofod** [comment]
Benchmarks for the 40 variables case. "Improved"-version is the one after changes by @andreasnoack .
results_mem
results_time

**Comment by pkofod** [comment]
Just a tiny note on the memory usage. The quoted memory usage is not the total amount of installed RAM you would need. Quite a few of the allocations are temporary.

**Comment by pkofod** [comment]
Regarding the discussion in #4 . I have received permission to merge this implementation into our package from the authors. I am not sure how to document it, but here is what they wrote.

Roger Koenker:
"Patrick,

Quoting your cited discussion:  Arrrrggh! I hate those licensing issues.

Me too.  The matlab version of this algorithm was something of an afterthought.
After I wrote the R/Fortran versions I thought it might be useful to have a version in Ox, so a student and I made a translation that was posted on the web, and at some point Paul Eilers made a translation of the Ox version to Matlab.  I
subsequently modified Paul's code and posted that.   Paul made some further
changes in 2004 that are reflected in the code you reference.  My R work is all GPL and I have no problem with you using this in the form that you now have but I'd like to copy Paul and this and make sure that he is also ok with it.

Roger"

and Eilers followed up:

Paul Eilers:
"Dear Patrick,

[... unrelated text..] Anyone can freely use this software, for academic and commercial purposes. I wish you good luck with your package.

Regards,

Paul Eilers"

So we should be good to go!

**Comment by andreasnoack** [comment]
The problem is that Koenker uses GPL which is not compatible with the BSD-3 licence used in this package. Right now, if included, this code will make the whole package GPL and, therefore, unusable for commercial purposes. You could try to ask if he is willing to accept that the Julia version gets BSD-3 licenced. Basically, the question is if he is okay with commercial use of the code.

**Comment by pkofod** [comment]
I will ask him to be 100% sure, but the GPL license he mentions is regarding his R package quantreg, not the code I used for this implementation. It is a matlab-implementation by Paul Eilers, who wrote (see the second quote) that academic as well as commercial purposes are fine.

**Comment by vincentarelbundock** [comment]
My understanding is that if you own the copyright to a software, you cannot rescind a license you have already given (i.e., I wouldn't be able to withdraw the BSD on any copy that people have already downloaded), but that you can also distribute it to other people under any license you desire. As such, the GPL is not constraining here, since the copyright owner has specifically given us permission to distribute our version under BSD.

Might be a good idea to include a note to the effect that permission from X and Y were given to Patrick via email on [DATE].

**Comment by pkofod** [comment]
Koenker is onboard with BSD-3

"Patrick,

[... unrelated  content...]  I'm fine with BSD-3.

Roger"

**Comment by pkofod** [comment]
By the way, I can provide the full-length mails if someone wants to see them, but they contain another, totally unrelated, conversation, which is why I put the [... unrelated content ...] tags :)

File: src/InteriorPoint.jl [hunk 1]
```diff
@@ -0,0 +1,135 @@
+# A Frisch-Newton algorithm as described in http://projecteuclid.org/euclid.ss/1030037960
+# qreg_ip_coef is a modified lp_fnm of Daniel Morillo & Roger Koenker
+# Translated from Ox to Matlab by Paul Eilers 1999
+# Modified by Roger Koenker 2000
+# More changes by Paul Eilers 2004
+# QuantileRegression.jl developers 2015
+#
+# The functions in this file is released under the BSD-3 license, although the original
```
Comment (pkofod) [inline src/InteriorPoint.jl]: I added a notice. (Also note: Travis running, yay!)

**Comment by pkofod** [comment]
I also added the .travis.yml-file, so now we have Travis running. Kind of OT, but I will add overalls, so we can be sure that the package is tested thoroughly (currently we have zero tests, but I will add them afterwards).

**Comment by vincentarelbundock** [comment]
Looks good. Thanks!

Why is travis failing? Also, merge at will, no need to wait if you think things are ready.

**Comment by pkofod** [comment]
It is v0.4 that failed, and I think it is for purely idiosyncratic reasons. If you look at the details, it failed before even loading Julia. I've made it run again. I tried this before when PRing in Optim.jl.

**Comment by pkofod** [comment]
Now there is a propor error. I'll fix it

**Comment by pkofod** [comment]
@andreasnoack Do you have any insight as to why qrfact! throws the following error in v0.4?

https://travis-ci.org/vincentarelbundock/QuantileRegression.jl/jobs/60241850

**Comment by andreasnoack** [comment]
Yes I have. We have changed the API here to make the function type stable so in 0.4 the syntax is `qrfact(A, Val{true})`. It is not as pretty, but it is much faster. In particular inside an algorithm like this. It is important that the check is at compile time and not at runtime because it would be unnecessarily slow. I think one way to do this without writing two versions of the whole function is to define a macro like
```jl
macro atleastversion(x, A)
    if VERSION > eval(x)
        return :(qrfact(A, Val{true}))
    else
        return :(qrfact(A, pivot = true))
    end
end
```
and then write line 64 as
```jl
AQtF = @atleastversion(VERSION, Q*X)
```

**Comment by pkofod** [comment]
Thanks, I will fix it, so it works for the cutting edge people running v0.4 :)

**Comment by pkofod** [comment]
Thanks @vincentarelbundock  and @andreasnoack  for feedback and assistance! This closes #4 .
